### PR TITLE
Fix a bug in point-in-polygon kernel: if the point is collinear with an edge, result is asserted false

### LINKED ISSUE #1103: [BUG]: wrong result for point_in_polygon
### Version

23.06.00a+39.g86e8912a

### On which installation method(s) does this occur?

Conda

### Describe the issue

For a specific Point and Polygon, shapely indicates that the point is located within the polygon, while `cuspatial.point_in_polygon` returns False.

Plotting the shapely geometries confirms that the point is indeed within the polygon:

image

### Minimum reproducible example

```python
point_wkt = "POINT (-118.352838 34.039809)"
polygon_wkt = "POLYGON ((-118.385129 34.019818, -118.384899 34.019909, -118.384456 34.020044, -118.3842872 34.020084, -118.384198 34.020113, -118.3839117 34.0203741, -118.3837825 34.0204828, -118.3833126 34.0206029, -118.3829196 34.0207809, -118.3828241 34.0208242, -118.3826741 34.0208921, -118.3826207 34.0209342, -118.382508 34.021023, -118.381227 34.021215, -118.3805455 34.0212971, -118.379666 34.021403, -118.3793023 34.0214163, -118.3791995 34.02142, -118.37835 34.021451, -118.3782882 34.0218768, -118.3782672 34.0220219, -118.3782402 34.0220617, -118.377987 34.022435, -118.3777751 34.0225414, -118.3776001 34.0226293, -118.377539 34.02266, -118.3774788 34.0227749, -118.377332 34.023055, -118.3772502 34.0231939, -118.3771932 34.0232906, -118.376801 34.023956, -118.376761 34.024281, -118.376596 34.02536, -118.3764983 34.025866, -118.376478 34.025971, -118.3766918 34.0261454, -118.37676 34.026201, -118.376732 34.026354, -118.376706 34.026497, -118.376685 34.0266145, -118.3766682 34.0266693, -118.3766538 34.0267199, -118.376563 34.027016, -118.376511 34.027154, -118.37645 34.027313, -118.376377 34.027495, -118.376175 34.028008, -118.37606 34.028321, -118.3759441 34.0284508, -118.375819 34.028591, -118.3756613 34.0291205, -118.3756356 34.0291453, -118.375501 34.029275, -118.375057 34.029738, -118.3747069 34.0297638, -118.374623 34.02977, -118.3740637 34.0299243, -118.3734637 34.0300898, -118.373271 34.030143, -118.373251 34.030133, -118.372723 34.029869, -118.3724157 34.0298179, -118.372344 34.029806, -118.372146 34.029844, -118.3720132 34.0298781, -118.3719368 34.0298675, -118.3718002 34.0298372, -118.371724 34.0301227, -118.371679 34.03026, -118.372046 34.03037, -118.3719957 34.0303797, -118.371922 34.030394, -118.371844 34.030428, -118.371809 34.03045, -118.371785 34.030466, -118.371703 34.030548, -118.371639 34.030621, -118.3715826 34.0307018, -118.37147 34.030863, -118.371241 34.031369, -118.3711981 34.0314631, -118.3711396 34.0315915, -118.371097 34.031685, -118.370941 34.032031, -118.3709122 34.0320983, -118.370825 34.032302, -118.370648 34.032305, -118.370576 34.032434, -118.3707606 34.0324453, -118.3706288 34.0327679, -118.370579 34.0328898, -118.3705566 34.0329445, -118.370533 34.033002, -118.3705227 34.0330252, -118.3704786 34.0331247, -118.3703622 34.033387, -118.3703301 34.0334592, -118.370316 34.033491, -118.3702636 34.0336061, -118.3701835 34.033782, -118.370153 34.033849, -118.369939 34.034286, -118.369876 34.034422, -118.369831 34.034515, -118.369735 34.034741, -118.369633 34.034947, -118.369577 34.035051, -118.369529 34.035143, -118.369475 34.035273, -118.369454 34.035327, -118.3694199 34.0354126, -118.369397 34.03547, -118.369342 34.035595, -118.3692974 34.0356978, -118.369257 34.035791, -118.369212 34.03588, -118.369181 34.035957, -118.369062 34.036204, -118.368952 34.036494, -118.3689244 34.0365642, -118.368877 34.036685, -118.368845 34.036763, -118.368807 34.036881, -118.368775 34.037002, -118.3687392 34.03711, -118.368718 34.037256, -118.368702 34.037414, -118.368701 34.037564, -118.3687032 34.0376476, -118.36872 34.037804, -118.368734 34.037906, -118.36877 34.038084, -118.368799 34.038181, -118.3688012 34.0381877, -118.368825 34.038259, -118.3688536 34.0383323, -118.368864 34.038359, -118.36891 34.038472, -118.368956 34.038564, -118.369026 34.038696, -118.369128 34.038878, -118.3692748 34.0391161, -118.3693521 34.0392416, -118.3694115 34.0393404, -118.3695638 34.0395936, -118.369691 34.039805, -118.369517 34.039901, -118.3694476 34.0399404, -118.368941 34.040228, -118.368276 34.040601, -118.367437 34.041076, -118.3671157 34.0412566, -118.366898 34.041379, -118.3666508 34.0415208, -118.366469 34.041625, -118.3660018 34.0418886, -118.365799 34.042003, -118.365578 34.0421184, -118.365326 34.04225, -118.365083 34.042374, -118.3649811 34.0424216, -118.364728 34.04254, -118.364728 34.042459, -118.364734 34.042367, -118.364753 34.042256, -118.364783 34.042151, -118.3648277 34.0420352, -118.3648616 34.0419526, -118.3649315 34.0417663, -118.365079 34.041373, -118.36551 34.040249, -118.3656914 34.0397865, -118.3657053 34.0397511, -118.365723 34.039706, -118.3657518 34.0396351, -118.3657849 34.0395534, -118.3652059 34.039809, -118.3651851 34.039809, -118.3650703 34.0398802, -118.3644071 34.0401835, -118.3641779 34.0402618, -118.36378 34.0403945, -118.3633157 34.0405392, -118.3631389 34.0405918, -118.3625579 34.040775, -118.3620521 34.0409344, -118.3620585 34.0408772, -118.3620819 34.0408005, -118.362176 34.040548, -118.3623833 34.0400135, -118.3624102 34.0399441, -118.3626307 34.0393526, -118.3628914 34.0386551, -118.3631013 34.0380587, -118.363128 34.037983, -118.3631682 34.0378566, -118.3631759 34.0378202, -118.3631763 34.0377838, -118.363168 34.037738, -118.3624366 34.0381085, -118.362416 34.038119, -118.361787 34.03844, -118.361246 34.038713, -118.3610647 34.0388045, -118.36083 34.038923, -118.3604144 34.0392315, -118.360297 34.039256, -118.3602456 34.0394469, -118.3602189 34.0395461, -118.360195 34.039627, -118.360159 34.039744, -118.3597884 34.040751, -118.359589 34.041293, -118.3594235 34.0417329, -118.358972 34.041879, -118.358646 34.041959, -118.358205 34.042049, -118.358317 34.041727, -118.3586216 34.040903, -118.358653 34.040818, -118.3587141 34.040654, -118.358783 34.0404689, -118.358895 34.040168, -118.358951 34.039896, -118.358968 34.039789, -118.3588882 34.0398302, -118.3588385 34.0398558, -118.3580224 34.0402771, -118.3579377 34.0403208, -118.357823 34.04038, -118.357772 34.040591, -118.357768 34.040631, -118.3577416 34.040749, -118.3577543 34.040666, -118.3577183 34.0405728, -118.357428 34.040697, -118.356408 34.041176, -118.3563734 34.0411268, -118.3562537 34.0411886, -118.3561417 34.0412413, -118.356021 34.0412886, -118.3559284 34.0413197, -118.3558051 34.0413614, -118.3556337 34.0414103, -118.3554926 34.0414375, -118.3553679 34.0414553, -118.3552566 34.0414708, -118.355164 34.0414775, -118.3550385 34.0414764, -118.3549696 34.0414364, -118.3549132 34.0414341, -118.3547456 34.0414241, -118.3546276 34.0414141, -118.3545029 34.0413986, -118.3543553 34.0413819, -118.3540992 34.0413486, -118.3539101 34.0413252, -118.3537505 34.0413064, -118.3535627 34.0412841, -118.353493 34.0412758, -118.3533379 34.0412544, -118.3532521 34.0412433, -118.353152 34.0412309, -118.3530603 34.0412199, -118.3528244 34.0411882, -118.3525896 34.0411566, -118.3524407 34.0411377, -118.3522758 34.0411164, -118.3521358 34.0410984, -118.3519207 34.0410721, -118.3517295 34.041044, -118.3515592 34.0410229, -118.3511892 34.0409886, -118.3511501 34.040985, -118.3510703 34.0410134, -118.3509154 34.040979, -118.3507469 34.0409589, -118.3505559 34.0409173, -118.3502372 34.0408773, -118.3498587 34.0408299, -118.349685 34.040808, -118.349497 34.0407718, -118.3487801 34.0406337, -118.348595 34.040598, -118.3483732 34.0405609, -118.34835 34.040557, -118.3481508 34.0405207, -118.348114 34.040514, -118.347831 34.040454, -118.3477897 34.0404448, -118.347517 34.040384, -118.3472506 34.0403286, -118.3468561 34.0402454, -118.3467519 34.0402266, -118.3466498 34.0402082, -118.3464249 34.0401567, -118.3463557 34.0401473, -118.346266 34.0401845, -118.345946 34.0401396, -118.3455882 34.0400599, -118.3455021 34.0399811, -118.345415 34.039949, -118.3452106 34.0399111, -118.34507 34.039885, -118.345001 34.0398714, -118.344898 34.039851, -118.344789 34.03983, -118.344686 34.039819, -118.344587 34.039819, -118.344465 34.039825, -118.3442455 34.0398178, -118.3441637 34.0398211, -118.3440775 34.0398209, -118.343948 34.039822, -118.3437228 34.039821, -118.343475 34.03982, -118.3432963 34.0398194, -118.3431769 34.0398232, -118.343097 34.039858, -118.34304 34.039884, -118.3429846 34.0398857, -118.3429001 34.0398835, -118.3428426 34.039882, -118.3427417 34.0398797, -118.3425743 34.0398779, -118.342395 34.039876, -118.3421008 34.0398778, -118.3416263 34.0398807, -118.341412 34.039882, -118.3411118 34.039884, -118.3410377 34.0398845, -118.340822 34.039886, -118.340381 34.039887, -118.340167 34.039884, -118.339785 34.03989, -118.3392199 34.0398911, -118.3387383 34.0398964, -118.3386263 34.0398909, -118.3385648 34.0398854, -118.3384807 34.0398366, -118.3384129 34.0398366, -118.3382595 34.0398373, -118.338026 34.0398377, -118.3379441 34.0398399, -118.337603 34.039843, -118.337344 34.039841, -118.337204 34.039842, -118.337075 34.039847, -118.336571 34.039845, -118.336284 34.039845, -118.33536 34.039853, -118.335007 34.03985, -118.33398 34.039854, -118.333688 34.039856, -118.33246 34.039849, -118.3318788 34.0398633, -118.3317244 34.0398642, -118.331411 34.039866, -118.331208 34.039872, -118.33114 34.039873, -118.331056 34.039869, -118.331088 34.039806, -118.3312021 34.0395994, -118.3312703 34.0394759, -118.3313964 34.0392476, -118.3315026 34.0390554, -118.331947 34.038277, -118.3325847 34.0371398, -118.3326607 34.0370064, -118.3327152 34.0369116, -118.3327967 34.0367659, -118.3328854 34.036621, -118.3330601 34.0363032, -118.333048 34.0362167, -118.3330529 34.0361735, -118.3332717 34.0357327, -118.3333152 34.0356331, -118.3333352 34.0355907, -118.333387 34.03549, -118.3335189 34.0352584, -118.33353 34.035239, -118.333591 34.035136, -118.333651 34.035041, -118.333709 34.034953, -118.3337797 34.0348236, -118.333884 34.034633, -118.333985 34.03444, -118.3340193 34.0343737, -118.3340887 34.0342394, -118.3341532 34.0341181, -118.3342295 34.033989, -118.3342701 34.0339278, -118.3344047 34.0338595, -118.334542 34.033599, -118.3346368 34.0334173, -118.3346614 34.0333702, -118.3347181 34.0332614, -118.334843 34.033022, -118.334899 34.032901, -118.3349224 34.0328381, -118.334935 34.032804, -118.33496 34.032733, -118.334987 34.032644, -118.335005 34.032577, -118.3350502 34.0323737, -118.3350567 34.0323417, -118.335069 34.032239, -118.335078 34.03216, -118.3350809 34.0321243, -118.335086 34.032063, -118.3350917 34.0318648, -118.3350972 34.0316722, -118.335099 34.031607, -118.3351031 34.0313639, -118.335107 34.031135, -118.3352105 34.0311347, -118.335318 34.031134, -118.3356388 34.0311304, -118.335767 34.031129, -118.336343 34.031125, -118.3363436 34.0308841, -118.3357651 34.0308871, -118.3355458 34.0308875, -118.3353354 34.030888, -118.3353155 34.030888, -118.3351354 34.0308887, -118.3351078 34.0308888, -118.3349105 34.0308898, -118.3345976 34.0308914, -118.3344126 34.0308924, -118.334411 34.031138, -118.333801 34.031146, -118.333796 34.029475, -118.334422 34.029464, -118.335103 34.029454, -118.335101 34.0290046, -118.3351002 34.0288196, -118.3350988 34.0285089, -118.3350973 34.0281665, -118.3350968 34.0280385, -118.3350959 34.0278282, -118.335095 34.027614, -118.3350947 34.0271789, -118.3350942 34.0264431, -118.3350941 34.0262137, -118.3350941 34.0261383, -118.335094 34.026032, -118.3350959 34.0258137, -118.3350979 34.0255753, -118.3350982 34.0253936, -118.3350986 34.0251072, -118.3350981 34.0250029, -118.335097 34.024786, -118.3350871 34.0242391, -118.335083 34.02401, -118.3350855 34.0237475, -118.3350867 34.0236214, -118.335088 34.0234803, -118.3350895 34.0233219, -118.33509 34.023271, -118.3350905 34.0232241, -118.3350913 34.0231575, -118.335097 34.022655, -118.335097 34.0225742, -118.3350982 34.0224813, -118.335098 34.0224, -118.3351125 34.02229, -118.3351218 34.0221734, -118.335134 34.0220155, -118.3351378 34.0219664, -118.3351498 34.0216465, -118.335146 34.021515, -118.3354065 34.021511, -118.3357636 34.0215149, -118.335773 34.021515, -118.3358619 34.0215202, -118.336319 34.021547, -118.336325 34.021436, -118.3361418 34.0214353, -118.3359594 34.0214347, -118.3358455 34.0214343, -118.335759 34.021434, -118.335672 34.0214332, -118.3354887 34.0214315, -118.3354109 34.0214309, -118.3351438 34.0214242, -118.3350162 34.0214218, -118.3348191 34.02141, -118.334686 34.021402, -118.3345628 34.0213825, -118.334411 34.021358, -118.3343824 34.0213519, -118.3342764 34.0213319, -118.3342295 34.0213247, -118.3341282 34.0212997, -118.333902 34.021242, -118.3337089 34.021197, -118.333456 34.021138, -118.3334404 34.0204575, -118.3334294 34.0194852, -118.3334284 34.0193926, -118.3334311 34.019346, -118.3334301 34.0192839, -118.3334353 34.0182107, -118.3334988 34.0182115, -118.333877 34.0182161, -118.3339669 34.0182172, -118.3340607 34.0182168, -118.334286 34.018216, -118.3343813 34.018217, -118.3345993 34.0182192, -118.3346669 34.0182199, -118.334682 34.01822, -118.3350591 34.0182088, -118.3351162 34.0182093, -118.3353554 34.0182142, -118.3356188 34.0182163, -118.335829 34.018218, -118.3359918 34.0182213, -118.336316 34.018228, -118.336323 34.017976, -118.336334 34.017606, -118.3363627 34.0172679, -118.33638 34.017065, -118.336398 34.016901, -118.336473 34.016512, -118.336592 34.016056, -118.336799 34.015431, -118.337033 34.014924, -118.337652 34.015153, -118.337948 34.015251, -118.338858 34.015564, -118.339207 34.015673, -118.339752 34.015885, -118.340188 34.016044, -118.34067 34.01621, -118.340718 34.016121, -118.340839 34.015916, -118.340973 34.015717, -118.341118 34.015523, -118.341274 34.015336, -118.3413334 34.0152704, -118.341342 34.015261, -118.3414655 34.0151282, -118.341596 34.014988, -118.342381 34.01549, -118.343171 34.015973, -118.3435748 34.0162247, -118.3437699 34.0163513, -118.3439389 34.0164502, -118.3446797 34.0169216, -118.3448525 34.0170316, -118.345799 34.017634, -118.345851 34.017578, -118.345915 34.017525, -118.345967 34.017484, -118.346028 34.017436, -118.3462333 34.0172451, -118.34626 34.017223, -118.3463021 34.0171892, -118.346331 34.017166, -118.346431 34.0171, -118.346535 34.017055, -118.346602 34.017031, -118.346676 34.017006, -118.34679 34.016973, -118.346891 34.016947, -118.346945 34.016939, -118.347003 34.016932, -118.347063 34.016928, -118.347141 34.016931, -118.3471807 34.0166886, -118.347191 34.016626, -118.347236 34.016352, -118.347412 34.016363, -118.34748 34.016373, -118.34753 34.016387, -118.347563 34.0164, -118.349647 34.017718, -118.349715 34.017747, -118.349767 34.017758, -118.349836 34.017761, -118.350442 34.01776, -118.350438 34.017965, -118.350439 34.018033, -118.350439 34.018277, -118.351068 34.0182805, -118.351152 34.018281, -118.352129 34.01828, -118.3524393 34.0182805, -118.352694 34.018281, -118.3531874 34.0182852, -118.353636 34.018289, -118.3541216 34.018292, -118.354284 34.018293, -118.3543913 34.0182959, -118.3545548 34.0183003, -118.3548077 34.0183071, -118.3550441 34.0183135, -118.355061 34.018314, -118.355577 34.018326, -118.355773 34.018323, -118.3561427 34.0183235, -118.3562782 34.0183237, -118.356556 34.018324, -118.356875 34.018326, -118.3571868 34.018326, -118.357954 34.018326, -118.3581463 34.0183216, -118.358438 34.018315, -118.358744 34.018294, -118.359151 34.018227, -118.359849 34.018088, -118.360138 34.018044, -118.360595 34.018016, -118.360865 34.018012, -118.361149 34.018033, -118.36157 34.018095, -118.362124 34.018212, -118.3623057 34.0182442, -118.362491 34.018277, -118.36279 34.018317, -118.363103 34.018337, -118.3632294 34.0183373, -118.364074 34.018339, -118.365034 34.018345, -118.366046 34.018356, -118.367245 34.018385, -118.367667 34.018426, -118.368152 34.018513, -118.368717 34.018687, -118.369713 34.019015, -118.370258 34.019202, -118.370816 34.019363, -118.371122 34.019422, -118.371759 34.019502, -118.3717819 34.019124, -118.371795 34.018908, -118.371822 34.018707, -118.371844 34.018644, -118.372012 34.018435, -118.372099 34.0183, -118.372521 34.018757, -118.372663 34.018807, -118.3729456 34.0187858, -118.373331 34.018757, -118.374132 34.0186768, -118.374163 34.01867, -118.374683 34.0186217, -118.3747325 34.0186358, -118.3749704 34.018593, -118.37515 34.018575, -118.37553 34.018544, -118.375877 34.018535, -118.375871 34.01808, -118.376741 34.018144, -118.3768397 34.0181332, -118.3769407 34.0181221, -118.3775863 34.0180513, -118.377659 34.0180406, -118.377731 34.01803, -118.377907 34.017962, -118.3779626 34.017941, -118.3780172 34.0179201, -118.378145 34.017872, -118.378983 34.018289, -118.37942 34.018318, -118.379945 34.018591, -118.381257 34.018323, -118.3817718 34.0183365, -118.3820641 34.0183442, -118.3822725 34.0185055, -118.3826503 34.0186534, -118.3828659 34.018573, -118.383118 34.018479, -118.3832993 34.0185894, -118.3843339 34.0191744, -118.384471 34.019242, -118.384186 34.019427, -118.383936 34.019608, -118.383748 34.019735, -118.383625 34.0197969, -118.383323 34.01995, -118.38311 34.0200607, -118.3829951 34.0201175, -118.3828177 34.0202052, -118.3827452 34.0202118, -118.3826835 34.0202029, -118.3826192 34.0201939, -118.3825843 34.0202006, -118.3825602 34.020214, -118.382528 34.0202452, -118.3825146 34.0202652, -118.3825065 34.0202985, -118.3825065 34.0203519, -118.3825119 34.020403, -118.383875 34.019901, -118.3840226 34.0198915, -118.3841676 34.0198822, -118.3842162 34.0198772, -118.3844553 34.0198527, -118.384696 34.019828, -118.384983 34.019727, -118.385074 34.019784, -118.385129 34.019818))"

import shapely

point = shapely.from_wkt(point_wkt)
polygon = shapely.from_wkt(polygon_wkt)
print(polygon.contains(point))
# -> True

import cuspatial
print(cuspatial.point_in_polygon(cuspatial.GeoSeries([point]), cuspatial.GeoSeries([polygon])))
# -> False
```

### Environment details

Installed from rapidsai-nightly channel on Colab

### PR DESCRIPTION
**Pull request by isVoid**
## Description

Fixes #1103, current algorithm tests if the test point is collinear with an edge of the polygon, the point is asserted to be on the edge. This is not true, because for a point to be on the edge, the point also needs to be within the range where the edge covers. Collinearity test only test if the point is covered by the line that the edge coincides.

This PR fixes this bug by adding additional tests to guarantee that the point is on an edge iff the point is collinear with the line where the edge coincides as well as the x coordinate of the point is within the closed range of the edge's x coordinates.

This PR also fixes an additional bug where the col-linearity flag is not reset after each iteration of a ring.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cuspatial/detail/algorithm/is_point_in_polygon.cuh [hunk 1]
```diff
@@ -39,25 +41,14 @@ namespace detail {
...
- * API.
  */
 template <typename T, class PolygonRef>
 __device__ inline bool is_point_in_polygon(vec_2d<T> const& test_point, PolygonRef const& polygon)
 {
   bool point_is_within = false;
-  bool is_colinear     = false;
+  bool point_on_edge   = false;
   for (auto ring : polygon) {
+    bool is_colinear  = false;
```
Comment (harrism) [inline cpp/include/cuspatial/detail/algorithm/is_point_in_polygon.cuh:51]: Only used below, no need to define this early.

```suggestion
```

**Review by harrism — APPROVED** [review]
Thanks for the quick fix!
